packages/ts-client/src/webrtc/StateManager.ts

```diff
@@ -0,0 +1,548 @@
+
+        this.webrtc.emit('trackAdded', ctx);
+      }
+    });
+  }
+
+  public onTracksRemoved(data: any) {
+    const endpointId = data.endpointId;
+    if (this.getEndpointId() === endpointId) return;
+    const trackIds = data.trackIds as string[];
+    trackIds.forEach((trackId) => {
+      const trackContext = this.trackIdToTrack.get(trackId)!;
+
+      this.eraseTrack(trackId, endpointId);
+
+      this.webrtc.emit('trackRemoved', trackContext);
+    });
+  }
+
+  public onSdpAnswer(data: any) {
+    this.midToTrackId = new Map(Object.entries(data.midToTrackId));
+
+    for (const trackId of Object.values(data.midToTrackId)) {
+      const track = this.localTrackIdToTrack.get(trackId as string);
+
+      // if is local track
+      if (track) {
+        track.negotiationStatus = 'done';
+
+        if (track.pendingMetadataUpdate) {
+          const mediaEvent = generateMediaEvent('updateTrackMetadata', {
+            trackId,
+            trackMetadata: track.metadata,
+          });
+          this.webrtc.sendMediaEvent(mediaEvent);
+        }
+
+        track.pendingMetadataUpdate = false;
+      }
+    }
+
+    this.onAnswer(data);
+  }
+
+  public onEndpointAdded(
+    endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata>,
+  ) {
+    if (endpoint.id === this.getEndpointId()) return;
+    endpoint.rawMetadata = endpoint.metadata;
+    try {
+      endpoint.metadataParsingError = undefined;
+      endpoint.metadata = this.endpointMetadataParser(endpoint.rawMetadata);
+    } catch (error) {
+      endpoint.metadataParsingError = error;
+      endpoint.metadata = undefined;
+    }
+    this.addEndpoint(endpoint);
+
+    this.webrtc.emit('endpointAdded', endpoint);
+  }
+
+  public onEndpointUpdated(data: any) {
+    if (this.getEndpointId() === data.id) return;
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
+      this.idToEndpoint.get(data.id)!;
+
+    try {
+      endpoint.metadata = this.endpointMetadataParser(data.metadata);
+      endpoint.metadataParsingError = undefined;
+    } catch (error) {
+      endpoint.metadata = undefined;
+      endpoint.metadataParsingError = error;
+    }
+    endpoint.rawMetadata = data.metadata;
+    this.addEndpoint(endpoint);
+
+    this.webrtc.emit('endpointUpdated', endpoint);
+  }
+
+  public onEndpointRemoved(data: any) {
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
+      this.idToEndpoint.get(data.id)!;
+    if (endpoint === undefined) return;
+
+    Array.from(endpoint.tracks.keys()).forEach((trackId) => {
+      this.webrtc.emit('trackRemoved', this.trackIdToTrack.get(trackId)!);
+    });
+
+    this.eraseEndpoint(endpoint);
+
+    this.webrtc.emit('endpointRemoved', endpoint);
+  }
+
+  public onTrackUpdated(data: any) {
+    if (this.getEndpointId() === data.endpointId) return;
+
+    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> =
```

Comment:
```suggestion
    const endpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata> | null =
```

(in previus funciton you are checking for undefined, but here for null. What is correct version?

Reply:
I will do it in future PRs